### PR TITLE
[WIP] catch requests from invalid thread

### PR DESCRIPTION
**Pull request by bfredl**

### CODE REVIEW

File: neovim/api/nvim.py [hunk 1]
```diff
@@ -126,6 +127,15 @@ def request(self, name, *args, **kwargs):
         present and True, a asynchronous notification is sent instead. This
         will never block, and the return value or error is ignored.
         """
+        if (self._session._loop_thread is not None and 
+                threading.current_thread() != self._session._loop_thread):
+
+            msg = ("request was make from a non-main thread:\n{}\n"
```
Comment (justinmk) [inline neovim/api/nvim.py]: "make" => "made". IMO could be reduced to "request from non-main thread:"

**Comment by phmarek** [comment]
```
request was make from a non-main thread:
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/plugin/script_host.py", line 226, in find_spec
    return PathFinder.find_spec(fullname, path or _get_paths(), target)
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/plugin/script_host.py", line 190, in _get_paths
    return discover_runtime_directories(nvim)
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/plugin/script_host.py", line 239, in discover_runtime_directories
    for path in nvim.list_runtime_paths():
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/api/nvim.py", line 251, in list_runtime_paths
    return self.request('nvim_list_runtime_paths')
```

And then some follow-ups:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "~/.config/nvim/bundle/YouCompleteMe/autoload/../python/ycm/client/base_request.py", line 144, in Session
    return cls.session
AttributeError: type object 'BaseRequest' has no attribute 'session'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "~/.config/nvim/bundle/YouCompleteMe/autoload/../python/ycm/client/base_request.py", line 216, in HandleServerException
    yield
  File "~/.config/nvim/bundle/YouCompleteMe/autoload/../python/ycm/client/ycmd_keepalive.py", line 50, in _ThreadMain
    BaseRequest.GetDataFromHandler( 'healthy' )
  File "~/.config/nvim/bundle/YouCompleteMe/autoload/../python/ycm/client/base_request.py", line 70, in GetDataFromHandler
    timeout ) )
  File "~/.config/nvim/bundle/YouCompleteMe/autoload/../python/ycm/client/base_request.py", line 110, in _TalkToHandlerAsync
    return BaseRequest.Session().get(
  File "~/.config/nvim/bundle/YouCompleteMe/autoload/../python/ycm/client/base_request.py", line 147, in Session
    from requests_futures.sessions import FuturesSession
  File "~/.config/nvim/bundle/YouCompleteMe/third_party/requests-futures/requests_futures/sessions.py", line 24, in <module>
    from requests import Session
  File "~/.config/nvim/bundle/YouCompleteMe/third_party/ycmd/third_party/requests/requests/__init__.py", line 53, in <module>
    from .packages.urllib3.contrib import pyopenssl
  File "~/.config/nvim/bundle/YouCompleteMe/third_party/ycmd/third_party/requests/requests/packages/urllib3/contrib/pyopenssl.py", line 49, in <module>
    from ndg.httpsclient.ssl_peer_verification import SUBJ_ALT_NAME_SUPPORT
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 954, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 896, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1147, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1121, in _get_spec
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/plugin/script_host.py", line 226, in find_spec
    return PathFinder.find_spec(fullname, path or _get_paths(), target)
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/plugin/script_host.py", line 190, in _get_paths
    return discover_runtime_directories(nvim)
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/plugin/script_host.py", line 239, in discover_runtime_directories
    for path in nvim.list_runtime_paths():
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/api/nvim.py", line 251, in list_runtime_paths
    return self.request('nvim_list_runtime_paths')
  File "/usr/lib/python3/dist-packages/neovim-0.1.14.dev0-py3.5.egg/neovim/api/nvim.py", line 137, in request
    raise NvimError("request was made from a non-main thread")
neovim.api.nvim.NvimError: request was made from a non-main thread
During handling of the above exception, another exception occurred:
...
```

**Comment by bfredl** [comment]
It seems that the import path hook is not thread safe, we'll need to either disallow path_hook imports on threads or use a cached copy of rtp.

**Comment by bfredl** [comment]
Please try latest commit

**Comment by phmarek** [comment]
Looks good so far.... when merging, please keep the first commit too, so that the problem can be more easily diagnosed later on.

Thank you!

**Pull request merged**